cloudfoundry-client-lib/src/main/java/org/cloudfoundry/client/lib/rest/CloudControllerRestClientImpl.java

```diff
@@ -1545,73 +1419,46 @@ public void bindDropletToApp(UUID dropletGuid, UUID applicationGuid) {
-            }, buildGuid);
-
-        return resourceMapper.mapResource(responseEntity.getBody(), CloudBuild.class);
+    private <R> List<R> doGetResources(String urlPath, Map<String, Object> urlVariables, Class<R> resourceClass) {
+        List<Map<String, Object>> resourceList = getAllResources(urlPath, urlVariables);
+        List<R> resources = new ArrayList<>();
+        for (Map<String, Object> resource : resourceList) {
+            if (resource != null) {
+                resources.add(resourceMapper.mapResource(resource, resourceClass));
+            }
+        }
+        return resources;
     }
 
-    @Override
-    public List<CloudBuild> getBuildsForApplication(UUID applicationGuid) {
-        Map<String, Object> urlVars = new HashMap<>();
-        urlVars.put("applicationGuid", applicationGuid);
-        return doGetResources("/v3/apps/{applicationGuid}/builds", urlVars, CloudBuild.class);
+    private List<Map<String, Object>> getAllResources(String urlPath, Map<String, Object> urlVars) {
+        Map<String, Object> responseMap = getResponseMap(urlPath, urlVars);
+        List<Map<String, Object>> allResources = new ArrayList<>();
+        List<Map<String, Object>> newResources = (List<Map<String, Object>>) responseMap.get("resources");
+        if (newResources != null && !newResources.isEmpty()) {
+            allResources.addAll(newResources);
+        }
+        addAllRemainingResources(responseMap, allResources);
+        return allResources;
     }
 
-    @Override
-    public Upload getUploadStatus(String uploadToken) {
-        CloudPackage cloudPackage = getCloudPackage(uploadToken);
-        ErrorDetails errorDetails = ImmutableErrorDetails.builder()
-            .description(cloudPackage.getData()
-                .getError())
-            .build();
-
-        return ImmutableUpload.builder()
-            .status(cloudPackage.getStatus())
-            .errorDetails(errorDetails)
-            .build();
+    private Map<String, Object> getResponseMap(String urlPath, Map<String, Object> urlVars) {
+        String response = getResponse(urlPath, urlVars);
+        return JsonUtil.convertJsonToMap(response);
     }
 
-    protected void extractUriInfo(Map<String, UUID> existingDomains, String uri, Map<String, String> uriInfo) {
-        URI newUri = URI.create(uri);
-        String host = newUri.getScheme() != null ? newUri.getHost() : newUri.getPath();
-
-        String[] hostAndDomain = host.split(DEFAULT_HOST_DOMAIN_SEPARATOR, 2);
-        if (hostAndDomain.length != 2) {
-            throw new IllegalArgumentException("Invalid URI " + uri + " -- host or domain is not specified");
-        }
+    private String getResponse(String urlPath, Map<String, Object> urlVars) {
+        return urlVars == null ? getRestTemplate().getForObject(getUrl(urlPath), String.class)
+            : getRestTemplate().getForObject(getUrl(urlPath), String.class, urlVars);
```

Comment:
I thought that the RestTeplate will not be used any more?

Reply:
We're still using it because of 3 missing features in the v3 client:
- Retrieval of service instance/binding parameters.
- Creation of applications from docker images.
- Retrieval of builds based on package ID.
The second is contributed and we're waiting for a release. The other ones will probably be done in a separate backlog.